### PR TITLE
Remove Kokkos_ENABLE_CUDA_LDG_INTRINSIC option

### PR DESCRIPTION
**Pull request by masterleinad**
Related to #5621. Since https://github.com/kokkos/kokkos/pull/2156, `KOKKOS_ENABLE_CUDA_LDG_INTRINSIC` is always enabled and, hence, `CudaTextureFetch` is unused.
This pull request proposes to eventually remove the option and `CudaTextureFetch` since it hasn't been used in more than 3 years.

### CODE REVIEW

**Review by dalg24 — COMMENTED** [review]
Please discuss why it is OK to drop the `KOKKOS_ENABLE_CUDA_LDG_INTRINSIC` macro

File: BUILD.md [hunk 1]
```diff
@@ -148,8 +148,9 @@ Options can be enabled by specifying `-DKokkos_ENABLE_X`.
     * Whether to activate experimental lambda features
     * BOOL Default: OFF
 * Kokkos_ENABLE_CUDA_LDG_INTRINSIC
+    * Deprecated since 4.0
     * Whether to use CUDA LDG intrinsics
-    * BOOL Default: OFF
+    * BOOL Default: ON
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply BUILD.md:153]: Yes, I was assuming that people if at all set `Kokkos_ENABLE_CUDA_LDG_INTRINSIC=ON` and not `Kokkos_ENABLE_CUDA_LDG_INTRINSIC=OFF` explicitly. Hence, I would only warn/error out of people set it to `ON` (and that's why the "real" default should be `OFF`). Any better suggestions?

File: Makefile.kokkos [hunk 2]
```diff
@@ -77,6 +77,7 @@ KOKKOS_INTERNAL_DISABLE_COMPLEX_ALIGN := $(call kokkos_has_string,$(KOKKOS_OPTIO
 KOKKOS_INTERNAL_DISABLE_DUALVIEW_MODIFY_CHECK := $(call kokkos_has_string,$(KOKKOS_OPTIONS),disable_dualview_modify_check)
 KOKKOS_INTERNAL_ENABLE_PROFILING_LOAD_PRINT := $(call kokkos_has_string,$(KOKKOS_OPTIONS),enable_profile_load_print)
 KOKKOS_INTERNAL_ENABLE_LARGE_MEM_TESTS := $(call kokkos_has_string,$(KOKKOS_OPTIONS),enable_large_mem_tests)
+# deprecated
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply Makefile.kokkos:80]: I find it better visible to put it on a separate line (and that's also what https://github.com/kokkos/kokkos/pull/5608 does).

File: cmake/KokkosCore_config.h.in [hunk 3]
```diff
@@ -36,7 +36,6 @@
 #cmakedefine KOKKOS_ENABLE_CUDA_UVM
 #cmakedefine KOKKOS_ENABLE_CUDA_LAMBDA
 #cmakedefine KOKKOS_ENABLE_CUDA_CONSTEXPR
-#cmakedefine KOKKOS_ENABLE_CUDA_LDG_INTRINSIC
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply cmake/KokkosCore_config.h.in:39]: I'm just keeping it now (912e82139c9d7ac645a85e95d59ef6419e51c0fd) to be conservative.

File: core/src/Cuda/Kokkos_Cuda_View.hpp [hunk 4]
```diff
@@ -263,28 +179,7 @@ class ViewDataHandle<
 
         record* const r = arg_tracker.template get_record<memory_space>();
 
-        if (0 == r) {
-          Kokkos::abort(
-              "Cuda const random access View using Cuda texture memory "
-              "requires "
-              "Kokkos to allocate the View's memory");
-        }
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_View.hpp:271]: I didn't change anything here. I just deleted the code in the obsolete branch (the display of the diff is a little unfortunate).

**Comment by dalg24** [comment]
> > Please discuss why it is OK to drop the KOKKOS_ENABLE_CUDA_LDG_INTRINSIC macro
> 
> I reverted it for it always to be defined (but not used).

Did you consider defining it only if deprecated code 4 is ON?  How likely is it that this macro is being used downstream?  Did you check in Trilinos?  etc.

**Comment by masterleinad** [comment]
> Did you consider defining it only if deprecated code 4 is ON? How likely is it that this macro is being used downstream? Did you check in Trilinos? etc.
 
There really is no good reason why the macro would be used downstream, though; it should be purely internal. `Trilinos` doesn't use it (except for the bundled `Kokkos` of course). My initial instinct was to remove it because of that but that might be strong. There is also `KOKKOS_OPT_RANGE_AGGRESSIVE_VECTORIZATION` which we should have removed by now. In view of that, it might be best to remove it conditionally.

Anyway, the more important question is, of course, if we are OK with removing the option (and with it the `CudaTextureFetch` branch) in the first place.

File: cmake/KokkosCore_config.h.in [hunk 5]
```diff
@@ -56,6 +55,10 @@
 #cmakedefine KOKKOS_ENABLE_AGGRESSIVE_VECTORIZATION
 #cmakedefine KOKKOS_ENABLE_IMPL_MDSPAN
 
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+#define KOKKOS_ENABLE_CUDA_LDG_INTRINSIC  // deprecated
+#endif
+
```
Comment (dalg24) [inline cmake/KokkosCore_config.h.in:61]: Belongs to `<Kokkos_Macros.hpp>`

Comment (dalg24) [reply cmake/KokkosCore_config.h.in:61]: Also we only want to define it if CUDA is enabled.

File: BUILD.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply BUILD.md:153]: The option needs to only be defined if deprecated code 4 is ON.  I am torn whether I would bother changing the documentation of what the default might be.  I am leaning towards no.

File: cmake/kokkos_enable_options.cmake [hunk 6]
```diff
@@ -141,3 +141,11 @@ ENDIF()
 IF (KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE AND BUILD_SHARED_LIBS)
   MESSAGE(FATAL_ERROR "Relocatable device code requires static libraries.")
 ENDIF()
+
+IF(KOKKOS_ENABLE_CUDA_LDG_INTRINSIC)
```
Comment (dalg24) [inline cmake/kokkos_enable_options.cmake:145]: You would do `KOKKOS_ -> Kokkos_` here if the option is conditionally declared.

File: cmake/kokkos_enable_options.cmake [hunk 7]
```diff
@@ -141,3 +141,11 @@ ENDIF()
...
+  IF(KOKKOS_ENABLE_DEPRECATED_CODE_4)
+    MESSAGE(DEPRECATION "Setting Kokkos_ENABLE_CUDA_LDG_INTRINSIC is deprecated. LDG intrinsics are always enabled.")
+  ELSE()
+    MESSAGE(FATAL_ERROR "Kokkos_ENABLE_DEPRECATED_CODE_4 must be set to use Kokkos_ENABLE_CUDA_LDG_INTRINSIC")
```
Comment (dalg24) [inline cmake/kokkos_enable_options.cmake:149]:
```suggestion
    MESSAGE(FATAL_ERROR "Kokkos_ENABLE_CUDA_LDG_INTRINSIC has been removed. LDG intrinsics are always enabled.")
```

**Review by ajpowelsnl — APPROVED** [review]

**Review by dalg24 — COMMENTED** [review]
Please confirm that you meant to define the option regardless of whether deprecated code 4 is enabled or not
https://github.com/kokkos/kokkos/blob/21d3e48e059646974d9ce367ceb184b894321766/cmake/kokkos_enable_options.cmake#L34

**Review by dalg24 — APPROVED** [review]
Will merge as is and we will do a follow up if necessary

**Pull request merged**